python/lsst/daf/butler/remote_butler/_remote_butler.py

```diff
@@ -179,6 +211,57 @@ def getURI(
         # Docstring inherited.
         raise NotImplementedError()
 
+    def get_dataset_type(self, name: str) -> DatasetType:
+        # In future implementation this should directly access the cache
+        # and only go to the server if the dataset type is not known.
+        path = f"dataset_type/{name}"
+        response = self._client.get(self._get_url(path))
+        response.raise_for_status()
+        return DatasetType.from_simple(SerializedDatasetType(**response.json()), universe=self.dimensions)
+
+    def get_dataset(self, id: DatasetId) -> DatasetRef | None:
+        path = f"dataset/{id}"
+        response = self._client.get(self._get_url(path))
+        if response.json() is None:
+            return None
+        return DatasetRef.from_simple(SerializedDatasetRef(**response.json()), universe=self.dimensions)
+
+    def find_dataset(
+        self,
+        datasetType: DatasetType | str,
+        dataId: DataId | None = None,
+        *,
+        collections: CollectionArgType | None = None,
+        timespan: Timespan | None = None,
+        datastore_records: bool = False,
+        **kwargs: Any,
+    ) -> DatasetRef | None:
+        if collections is None:
+            if not self.collections:
+                raise NoDefaultCollectionError(
+                    "No collections provided to find_dataset, and no defaults from butler construction."
+                )
+            collections = self.collections
+        # Temporary hack. Assume strings for collections. In future
+        # want to construct CollectionWildcard and filter it through collection
+        # cache to generate list of collection names.
```

Comment:
Working on caching ticket now, I would probably prefer to not have a complete list of collections on every client in the long term.

Reply:
Interesting. I thought we said we would last week. I thought @TallJimbo wanted the client to know all possible collections when it was building the query object before sending it to the server. The caching story between client and server is a bit tricky.

Reply:
Yes, I know, but we may need to rethink that part, otherwise we may have scaling issue. I looked at Butler initialization time today, the simplest `butler query-data-ids --collections=HSC/raw/all /repo/main exposure` takes 10 seconds, large fraction of that comes from pre-loading of collections cache. We have 300k collections today, and it will only grow with time. There are ways to reduce that probably, but in my ideal world it would be better if we cache almost nothing on client side. 🙂

Reply:
I think having the caches in the client has worked well for us so far (i.e. with the server just being the DB), and I like the idea of being able to optimize server calls away entirely in the client sometimes.  And I do think caching on the server is a lot harder, since we can't just punt cache misses to the user there.

But I do hesitate a bit because I'm not sure if "cache all the collections the user can see" continues to be in the "small enough to be irrelevant" category it is now after DR1, both in terms of downloading the cache and holding it in memory.  And I have no idea if my "do more on the client to keep the server load light" intuition is better than @dhirving 's "do nothing on the client to keep version management easier" intuition.

Reply:
Downloading a million collection records just to figure out that pattern does not match any of them looks like a lot of overhead. It is of course about the right balance between what we do on server vs client, we need to find something that works and scales.

Reply:
Ok, I'm sold on not trying to aggressively fetch all collections the user can access.  It's the dataset type definitions, chained collection members, and collection summaries that are needed for query contruction, and we don't need all of any of those.

I think what may break if we drop the complete collection-name cache is our regex/glob collection query implementation - we'll need to start translating/limiting those patterns to something that can be run inside the DB.  Yes, we could do a full cache in the `RemoteButler` server, but this scaling problem will hit `DirectButler`, too.

I'll do a brain dump on Slack later today about the places we use the caches in the query system so somebody other than me can do some conceptual design work on what it should look like.  I think we need to have a better sense of where this is all going, even if we don't decide to switch fully from the current approach yet, but I want to stay off of at least this part of the critical path.